silkworm/db/datastore/snapshots/basic_queries.hpp

```diff
@@ -92,19 +93,27 @@ struct FindByHashSegmentQuery : public BasicSegmentQuery<TDecoder, segment_names
     using BasicSegmentQuery<TDecoder, segment_names>::BasicSegmentQuery;
 
     std::optional<decltype(TDecoder::value)> exec(const Hash& hash) {
-        auto offset = this->index_.lookup_by_key(hash);
-        if (!offset) {
+        auto result = exec_with_timestamp(hash);
+        if (!result) return std::nullopt;
+        return result->first;
+    }
+
+    std::optional<std::pair<decltype(TDecoder::value), datastore::Timestamp>> exec_with_timestamp(const Hash& hash) {
```

Comment:
```suggestion
    std::optional<Result> exec(const Hash& hash) {
```

I suggest to rename this method to `exec` (and remove original exec), because like that it will become accessible to combinators that execute queries over a range of bundles (such as FindMapQuery).
The convention is that each segment query has a single exec method (maybe we should switch to `operator()` to make it mandatory).

This query is used by HeaderFindByHashQuery which is used in a single method - DataModel::read_header_from_snapshot, so a little modification there will be needed.

Also I suggest to define an inner Result struct containing value and timestamp instead of using std::pair. This improves readability on the call sites. We do this in a few places, e.g. see DomainGetLatestQuery in snapshots.

Reply:
I agree with the suggested change, i.e. `exec` instead of `exec_with_timestamp` but not with the switch to `operator()`, which in any case would not prevent having multiple methods through overloading.
The inner Result struct is really a plus for readability, thanks for the suggestion.

I'm gonna do this change into this PR.